### PR TITLE
HELM-214/HELM-222: Publish build artifacts with CircleCI to Cloudsmith

### PR DESCRIPTION
**Pull request by indigo423**
# Pull Request Check Sheet

* [x] Have you read and followed our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [x] Have you [made an issue in the OpenNMS issue tracker](https://issues.opennms.org)?<br>If so, you should:
  1. update the title of this PR to be of the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the JIRA link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
* [x] Have you made a comment in that issue which points back to this PR?
* [x] Have you updated the JIRA link at the bottom of this comment to link to your issue?
* [ ] If this is a new feature, is there documentation?
* [ ] If this is a new feature or substantial change, are there tests that cover it?

Created two build jobs to build DEB and RPM packages. Added the Cloudsmith orb to publish it to our [repository](https://cloudsmith.io/~opennms/repos/common/packages/). The Cloudsmith ORB requires the CLOUDSMITH_API_KEY environment variable to be set on the project.

## Reviewer hint

We create RPM and DEB packages in the `dist/packages` directory. In the `make-deb` and `make-rpm`, I've just set to store everything in the dist/packages dir see:

```yaml
      - store_artifacts:
          path: ./dist/packages
```
There is no wildcard support in the store_artifacts directive and I didn't want to mess around the makerpm and makedeb scripts.

# External References

* JIRA (Issue Tracker): https://issues.opennms.org/browse/HELM-214
* JIRA (Issue Tracker): https://issues.opennms.org/browse/HELM-222

### CODE REVIEW

**Comment by RangerRick** [comment]
Should it be a hard requirement that we support signing before we merge this for real?

**Comment by indigo423** [comment]
> Should it be a hard requirement that we support signing before we merge this for real?

absolutely not. We can add this here as well.

**Comment by indigo423** [comment]
@RangerRick I've made an issue for the signing in Helm ([HELM-222](https://issues.opennms.org/browse/HELM-222)). I've seen you created an issue for the signing orb, I've linked the issue as a dependency.

**Comment by indigo423** [comment]
@RangerRick no objections to be the test 🐰 for your orb :)

File: .circleci/config.yml [hunk 1]
```diff
@@ -238,6 +276,32 @@ jobs:
...
+          package-format: rpm
+          package-distribution: any-distro/any-version
+          package-path: dist/packages/*.rpm
+
+  publish-deb:
+    executor: cloudsmith/default
+    steps:
+      - attach_workspace:
+          at: ~/
+      - cloudsmith/ensure-api-key
```
Comment (j-white) [inline .circleci/config.yml]: We could consider using CircleCI contexts to populate the env. variables so we don't need to repeat them in every project.

File: .circleci/config.yml [hunk 2]
```diff
@@ -270,6 +334,18 @@ workflows:
           filters:
             tags:
               only: /^v.*/
+      - make-rpm:
```
Comment (j-white) [inline .circleci/config.yml]: What's the intended logic with the RPM/Deb building/publishing?

Build packages in all branches, but only push for tags starting with `v` and `master`?

Comment (indigo423) [reply .circleci/config.yml]: @j-white @RangerRick We do releases right now manually we haven't really talked about how we want to have the release logic. I would suggest following the approach from Alec where we publish release artifacts on git version tags prefixed with `v`, develop and release branches:

```yaml
filters:
  tags:
    only: /^v.*/
      branches:
        only:
          - develop
          - /release-.*/
```
It gives us a little bit more flexibility. We could do also something more restrictive and just publish release artifacts on git version tags:

```yaml
filters:
  tags:
    only: /^v.*/
      branches:
        ignore: /.*/
```
I'm open for one way or another. We can decide this here in the comments and fix the PR. I'll extract the release workflow from this conversation and we can it to the docs.

**Comment by indigo423** [comment]
@j-white @RangerRick I was able to get the signing with your orb to work and everything worked as expected. Do we want to decide in HELM-214 what kind of release workflow we want to have or should we go ahead and move it to a dedicated issue?

**Comment by j-white** [comment]
> @j-white @RangerRick I was able to get the signing with your orb to work and everything worked as expected. Do we want to decide in HELM-214 what kind of release workflow we want to have or should we go ahead and move it to a dedicated issue?

I think we should avoid pushing RPMs for every branch initially - maybe just develop for now and then we can refine it?

**Comment by RangerRick** [comment]
yeah, I would only do develop for now... we don't really have support for "branch" packages in cloudsmith anyway

**Comment by indigo423** [comment]
> yeah, I would only do develop for now... we don't really have support for "branch" packages in cloudsmith anyway

@j-white @RangerRick filter for publish-rpm and publish-deb is just set to develop. Should be good then so we can move forward and merge it. Any objections?

**Pull request merged**